cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java

```diff
@@ -0,0 +1,179 @@
+    long startTime = RunIds.getTime(RunIds.fromString(programRunId.getRun()), TimeUnit.MILLISECONDS);
+    if (startTime == -1) {
+      // If RunId is not time-based, use current time as start time
+      startTime = System.currentTimeMillis();
+    }
+    ImmutableMap.Builder properties = ImmutableMap.<String, String>builder()
+      .put(ProgramOptionConstants.START_TIME, String.valueOf(startTime))
+      .put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.STARTING.name())
+      .put(ProgramOptionConstants.USER_OVERRIDES, GSON.toJson(programOptions.getUserArguments().asMap()))
+      .put(ProgramOptionConstants.SYSTEM_OVERRIDES, GSON.toJson(programOptions.getArguments().asMap()));
+    if (twillRunId != null) {
+      properties.put(ProgramOptionConstants.TWILL_RUN_ID, twillRunId);
+    }
+    publish(programRunId, properties);
+  }
+
+  @Override
+  public void running(ProgramRunId programRunId, @Nullable String twillRunId) {
+    ImmutableMap.Builder properties = ImmutableMap.<String, String>builder()
+      .put(ProgramOptionConstants.LOGICAL_START_TIME, String.valueOf(System.currentTimeMillis()))
+      .put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.RUNNING.name());
+    if (twillRunId != null) {
+      properties.put(ProgramOptionConstants.TWILL_RUN_ID, twillRunId);
+    }
+    publish(programRunId, properties);
+  }
+
+  @Override
+  public void completed(ProgramRunId programRunId) {
+    stop(programRunId, ProgramRunStatus.COMPLETED, null);
+  }
+
+  @Override
+  public void killed(ProgramRunId programRunId) {
+    stop(programRunId, ProgramRunStatus.KILLED, null);
+  }
+
+  @Override
+  public void error(ProgramRunId programRunId, Throwable failureCause) {
+    stop(programRunId, ProgramRunStatus.FAILED, failureCause);
+  }
+
+  @Override
+  public void suspend(ProgramRunId programRunId) {
+    publish(
+      programRunId,
+      ImmutableMap.<String, String>builder()
+        .put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.SUSPENDED.name())
+    );
+  }
+
+  @Override
+  public void resume(ProgramRunId programRunId) {
+    publish(
+      programRunId,
+      ImmutableMap.<String, String>builder()
+        .put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.RESUMING.name())
+    );
+  }
+
+  private void stop(ProgramRunId programRunId, ProgramRunStatus runStatus, @Nullable Throwable cause) {
+    ImmutableMap.Builder properties = ImmutableMap.<String, String>builder()
+      .put(ProgramOptionConstants.END_TIME, String.valueOf(System.currentTimeMillis()))
+      .put(ProgramOptionConstants.PROGRAM_STATUS, runStatus.name());
+    if (cause != null) {
+      properties.put(ProgramOptionConstants.PROGRAM_ERROR, GSON.toJson(new BasicThrowable(cause)));
+    }
+    publish(programRunId, properties);
+  }
+
+  private void publish(ProgramRunId programRunId, ImmutableMap.Builder<String, String> properties) {
```

Comment:
it seems odd to just pass the programRunId to put in the properties. Better to have this method accept only properties which need to be send to the Notification ctor.